rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,337 @@
+  broadcast(msgType MsgType, rawData []byte)
+  validate(validation ValidationHandle, status ValidationStatus)
+
+Helper -> Daemon
+  handleRequest(requestId RequestId, from Peer, rawData []byte)
+  handleResponse(requestId RequestId, validation ValidationHandle, rawData []byte)
+  handleGossip(from Peer, validation ValidationHandle, rawData []byte)
+  stats(stats Stats)
+```
+
+```go
+// == The `Config` struct is sent with the `init` message at the start of the protocol.
+
+// The following old fields have been completely removed:
+//   - `metricsPort` (moving over to push-based stats syncing, where we will sync any metrics we want to expose)
+//   - `unsafeNotTrustIp` (only used as a hack in old integration test framework; having it makes p2p code harder to reason about)
+//   - `gaterConfig` (moving towards more abstracted interface in which Go manages gating state data)
+type Config struct {
+  networkId           string      // unique network identifier
+  privateKey          string      // libp2p id private key
+  stateDirectory      string      // directory to store state in (peerstore and dht will be stored/loaded from here)
+  listenOn            []Multiaddr // interfaces we listen on
+  externalAddr        Multiaddr   // interface we advertise for other nodes to connect to
+  floodGossip         bool        // enables gossip flooding (should only be turned on for protected nodes hidden behind a sentry node)
+  directPeers         []Multiaddr // forces the node to maintain connections with peers in this list (typically only used for sentry node setups and other specific networking scenarios; these peers are automatically trustlisted)
+  seedPeers           []Multiaddr // list of seed peers to connect to initially (seeds are automatically trustlisted)
+  maxConnections      int         // maximum number of connections allowed before the connection manager begins trimming open connections
+  validationQueueSize int         // size of the queue of active pending validation messages
+  // TODO: peerExchange bool vs minaPeerExchange bool (seems like at least one of these should be deprecated)
+  //   - peerExchange == enable libp2p's concept of peer exchange in the pubsub options
+  //   - minaPeerExchange == write random peers to connecting peers
+}
+
+// == The `Stats` struct is sent on an interval via the `stats` message.
+// == It contains metrics and statistics relevant to the helper process,
+// == to be further exposed by the daemon as prometheus metrics.
+
+type Stats struct {
+  storedPeerCount       int
+  connectedPeerCount    int
+  messageMinSize        float64
+  messageMaxSize        float64
+  messageAvgSize        float64
+  latencyMin            float64
+  latencyMax            float64
+  latencyAvg            float64
+  totalBandwidthIn      float64
+  totalBandwidthOut     float64
+  totalBandwidthRateIn  float64
+  totalBandwidthRateOut float64
+}
```

Comment:
you could create a `MinMaxAvg` struct and make `latency MinMaxAvg` and `message MinMaxAvg` (likewise with moving totalBandwidth* to another struct).

I think this could help prevent us from making a stupid mistake just due to misreading field names